app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt

```diff
@@ -18,141 +18,128 @@ package com.example.android.architecture.blueprints.todoapp.data
 
 import com.example.android.architecture.blueprints.todoapp.data.source.local.TasksDao
 import com.example.android.architecture.blueprints.todoapp.data.source.network.NetworkDataSource
-import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.map
-import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
 
 /**
  * Default implementation of [TasksRepository]. Single entry point for managing tasks' data.
+ *
+ * @param tasksNetworkDataSource - The network data source
+ * @param tasksDao - The local data source
+ * @param coroutineDispatcher - The dispatcher to be used for long running or complex operations,
+ * such as network calls or mapping many models. This is important to avoid blocking the calling
+ * thread.
  */
 class DefaultTasksRepository(
     private val tasksNetworkDataSource: NetworkDataSource,
     private val tasksDao: TasksDao,
+    private val coroutineDispatcher: CoroutineDispatcher = Dispatchers.Default
 ) : TasksRepository {
 
     override suspend fun createTask(title: String, description: String): Task {
         val task = Task(title = title, description = description)
-
-        coroutineScope {
-            launch { tasksNetworkDataSource.saveTask(task.toNetworkModel()) }
-            launch {
-                tasksDao.insertTask(task.toLocalModel())
-            }
-        }
+        tasksDao.insertTask(task.toLocalModel())
+        saveTasksToNetwork()
         return task
     }
 
     override suspend fun updateTask(taskId: String, title: String, description: String) {
-
         val task = getTask(taskId)?.copy(
             title = title,
             description = description
         ) ?: throw Exception("Task (id $taskId) not found")
 
-        coroutineScope {
-            launch { tasksNetworkDataSource.saveTask(task.toNetworkModel()) }
-            launch {
-                tasksDao.insertTask(task.toLocalModel())
-            }
-        }
+        tasksDao.insertTask(task.toLocalModel())
+        saveTasksToNetwork()
     }
 
     override suspend fun getTasks(forceUpdate: Boolean): List<Task> {
         if (forceUpdate) {
-            updateTasksFromRemoteDataSource()
+            loadTasksFromNetwork()
+        }
+        return withContext(coroutineDispatcher) {
+            tasksDao.getTasks().toExternalModels()
         }
-        return tasksDao.getTasks().map { it.toExternalModel() }
     }
 
     override suspend fun refreshTasks() {
-        updateTasksFromRemoteDataSource()
+        loadTasksFromNetwork()
     }
 
     override fun getTasksStream(): Flow<List<Task>> {
         return tasksDao.observeTasks().map { tasks ->
-            tasks.map { task ->
-                task.toExternalModel()
+            withContext(coroutineDispatcher) {
```

Comment:
Why do we need to change the context? `tasks.toExternalModels()` seems synchronous

Reply:
For exactly that reason, it is synchronous and could therefore block the calling thread indefinitely if it was, say, mapping a million tasks